### PR TITLE
Shooter subsystem

### LINKED ISSUE #14: Shooter Subsystem
the shooter subsystem contains a shooter motor and angle adjustment

required features are:
- dynamic driving gear ratios
- state space control
- current limits on motors
- functional input for angle and speed based off of position on field, f(x,y) -> (theta, v)
- logging of all relevant outputs, notable shooter speed (m/s eject velocity), shooter angle, shooter state, whether the shooter is "at target" ie spun up and at the correct angle for its input position

### PR DESCRIPTION
**Pull request by icai0**

### CODE REVIEW

File: constants.py [hunk 1]
```diff
@@ -685,3 +686,56 @@
 
 kVelocityControlMotorType = DCMotor.falcon500()
 kVelocityControlkV = 0.01
+
+kAngleMotorRatio = (64 / 16) * (60 / 18)
```
Comment (lmaxwell24) [inline constants.py:724]: assuming taken from cad

File: constants.py [hunk 2]
```diff
@@ -685,3 +686,56 @@
...
+kShootingMotorRatio = 24 / 36
+
+# change numbers later
+
+kAngleMotorCANId = 99
```
Comment (lmaxwell24) [inline constants.py]: even for "change later" keep below 63 / pick a number that would work

File: constants.py [hunk 3]
```diff
@@ -685,3 +686,56 @@
...
+kAngleMotorName = "ShooterAngleMotor"
+kAngleMotorPGain = 0.1
+kAngleMotorIGain = 0
+kAngleMotorDGain = 0
+kAngleMotorInverted = False
+
+kLeftShootingMotorCANId = 999
+kLeftShootingMotorName = "LeftShootingMotor"
+kLeftShootingMotorPIDSlot = 0
+kLeftShootingMotorPGain = 1
```
Comment (lmaxwell24) [inline constants.py]: oh god this would go way too fast
do some kV action for this motor
kV is in rps / V, so take max free speed and divide by voltage for that free speed

File: constants.py [hunk 4]
```diff
@@ -685,3 +686,56 @@
...
+kRightShootingMotorDGain = 0
+kRightShootingMotorInverted = False
+
+kAngleMotorMappingFunction = lambda x, y: x * y
+kLeftShootingMotorMappingFunction = lambda x, y: x * y
+kRightShootingMotorMappingFunction = lambda x, y: x * y
+kRobotAngleMappingFunction = lambda x, y: x * y
+
+kShooterAngleEncoderCANId = 1
+kShooterAngleMotorOffset = Rotation2d.fromDegrees(10.207848)
```
Comment (lmaxwell24) [inline constants.py]: whats defined as angle 0?
also is this number random? if so indicate must be changed

File: constants.py [hunk 5]
```diff
@@ -685,3 +686,56 @@
...
+
+kShooterAngleKey = "shooterAngle"
+kLeftShootingMotorSpeedKey = "leftShootingMotorRPM"
+kRightShootingMotorSpeedKey = "rightShootingMotorRPM"
+
+kShooterAngleTolerance = Rotation2d.fromDegrees(1)
```
Comment (lmaxwell24) [inline constants.py]: we showed our tolerance has to be 0.5 degrees

File: constants.py [hunk 6]
```diff
@@ -685,3 +686,56 @@
...
+kShooterSpeedTolerance = 100
```
Comment (lmaxwell24) [inline constants.py:770]: units?

File: constants.py [hunk 7]
```diff
@@ -685,3 +686,56 @@
...
+
+kShooterAngleOnTargetKey = "angleOnTarget"
+kLeftShootingMotorOnTargetKey = "leftMotorOnTarget"
+kRightShootingMotorOnTargetKey = "rightMotorOnTarget"
+kRobotAngleOnTargetKey = "robotAngleOnTarget"
+kReadyToShoot = "fireInTheHole"
```
Comment (lmaxwell24) [inline constants.py]: best name

File: constants.py [hunk 8]
```diff
@@ -685,3 +686,56 @@
...
+
+kShooterMaxAngle = Rotation2d.fromDegrees(53.820315)
```
Comment (lmaxwell24) [inline constants.py]: is there supposed to be an accompanying min angle?

File: subsystems/shootersubsystem.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        self.shooterEncoder = CTREEncoder(
+            constants.kShooterAngleEncoderCANId,
+            constants.kShooterAngleMotorOffset.degrees()
+            / constants.kDegeersPerRevolution,
+        )
+
+        self.leftShootingMotor.setSmartCurrentLimit(
+            constants.kShootingMotorCurrentLimit
+        )
+        self.rightShootingMotor.setSmartCurrentLimit(
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py:50]: likely should also have a smart current limit on the pivot

File: subsystems/shootersubsystem.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+            constants.kShooterAngleMotorOffset.degrees()
+            / constants.kDegeersPerRevolution
+            * constants.kShootingMotorRatio
+        )
+
+        self.targetAngle = Rotation2d()
+        self.leftTargetSpeed = 0
+        self.rightTargetSpeed = 0
+
+        # speeds are in RPM
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py]: keep it in radians per second in prog for SI sake

File: subsystems/shootersubsystem.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
+    def setShooterAngle(self, angle: Rotation2d) -> None:
+        self.targetAngle = min(max(Rotation2d(0), angle), constants.kShooterMaxAngle)
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py]: what is angle 0 in this context? suggested position is horizontal with the bot frame

File: subsystems/shootersubsystem.py [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+
+        self.angleMotor.set(
+            Talon.ControlMode.Position,
+            self.targetAngle.degrees()
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py]: nitpick but use radians :)
internally it does save some cycles since Rotation2ds are stored in radians

File: subsystems/shootersubsystem.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+            / constants.kDegeersPerRevolution
+            * constants.kAngleMotorRatio,
+        )
+
+    def setLeftShootingMotorSpeed(self, speed: int) -> None:
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py]: include units of input speed here
NEO expects it in revolutions per second, make sure its indicated as such

File: subsystems/shootersubsystem.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+
+    def setRightShootingMotorSpeed(self, speed: int) -> None:
+        self.rightTargetSpeed = speed
+        self.rightShootingMotor.set(
+            NEOBrushless.ControlMode.Velocity,
+            speed * constants.kShootingMotorRatio,
+        )
+
+    def getShooterAngle(self) -> Rotation2d:
+        return Rotation2d.fromDegrees(
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py]: radianssss

File: subsystems/shootersubsystem.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+            self.angleMotor.get(Talon.ControlMode.Position)
+            * constants.kAngleMotorRatio
+            * constants.kDegeersPerRevolution
+        )
+
+    def getLeftShooterSpeed(self) -> int:
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py:119]: comment for units

File: subsystems/shootersubsystem.py [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+
+    def angleOnTarget(self) -> bool:
+        return (
+            abs(self.targetAngle - self.getShooterAngle())
+            < constants.kShooterAngleTolerance
+        )
+
+    def leftMotorSpeedOnTarget(self) -> bool:
+        return (
+            abs(self.leftTargetSpeed - self.getLeftShooterSpeed)
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py]: should this function get called?

File: subsystems/shootersubsystem.py [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+            < constants.kShooterSpeedTolerance
+        )
+
+    def rightMotorSpeedOnTarget(self) -> bool:
+        return (
+            abs(self.rightTargetSpeed - self.getRightShooterSpeed)
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py]: rightTargetSpeed, nor leftTargetSpeed is never set post-init

**Review by lmaxwell24 — CHANGES_REQUESTED** [review]
also needs to put all the motors into motorSimulator
I'm working on adding Spark FLEXs to that lib, pretty far along with it rn so should be soon

needs to add shootersubsystem to robotcontainer.py as well
please also include some commands for debug control via smartdashboard

File: constants.py (hunk 4 above)
Comment (icai0) [reply constants.py]: angle from horizontal, from cad

File: constants.py (hunk 6 above)
Comment (icai0) [reply constants.py:770]: rpm

File: constants.py (hunk 8 above)
Comment (icai0) [reply constants.py]: it's the angle from min angle, which becomes 0 after applying the offset

File: constants.py (hunk 3 above)
Comment (icai0) [reply constants.py]: hopefully I did this right but please check over it

File: subsystems/shootersubsystem.py (hunk 10 above)
Comment (icai0) [reply subsystems/shootersubsystem.py]: velocity is in rpm

File: subsystems/shootersubsystem.py (hunk 13 above)
Comment (icai0) [reply subsystems/shootersubsystem.py]: rpm

File: subsystems/shootersubsystem.py (hunk 17 above)
Comment (icai0) [reply subsystems/shootersubsystem.py]: they are set in setLeftShootingMotorSpeed and right

File: subsystems/shootersubsystem.py (hunk 16 above)
Comment (icai0) [reply subsystems/shootersubsystem.py]: I don't really understand this question tbh

File: commands/shooter/decreaseleftmotorspeed.py [hunk 18]
```diff
@@ -0,0 +1,26 @@
+from commands2 import Command
+from wpilib import SmartDashboard
+
+import constants
+
+
+class DecreaseLeftMotorSpeed(Command):
+    def __init__(self) -> None:
+        Command.__init__(self)
+        self.setName(__class__.__name__)
```
Comment (lmaxwell24) [inline commands/shooter/decreaseleftmotorspeed.py:10]: I would somewhere also initialize this fudge key so that its set *before* its used

File: commands/shooter/shootermanualmode.py [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+
+class ShooterManualMode(Command):
+    def __init__(self, shooter: ShooterSubsystem) -> None:
+        Command.__init__(self)
+        self.setName(__class__.__name__)
+        self.shooter = shooter
+        self.addRequirements(self.shooter)
+
+    def execute(self) -> None:
+        self.shooter.setLeftShootingMotorSpeed(
```
Comment (lmaxwell24) [inline commands/shooter/shootermanualmode.py:19]: initialize these values before usage (add a SmartDashboard.putNumber() in __init__)

File: util/simneo.py [hunk 20]
```diff
@@ -49,6 +62,8 @@ def __init__(
         if not revCheckError("setD", self.controller.setD(dGain, pidSlot)):
             return
 
+        self.controller.ArbFFUnits.kVoltage = kV
```
Comment (lmaxwell24) [inline util/simneo.py]: see comments on discord about ff as a kv

File: subsystems/shootersubsystem.py [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+
+        self.leftShootingMotor.setSmartCurrentLimit(
+            constants.kShootingMotorCurrentLimit
+        )
+        self.rightShootingMotor.setSmartCurrentLimit(
+            constants.kShootingMotorCurrentLimit
+        )
+
+        self.angleMotor.setCurrentLimit(
+            CurrentLimitsConfigs().with_stator_current_limit(
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py]: I'd use a supply current limit here instead

File: subsystems/shootersubsystem.py [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+        self.targetAngle = Rotation2d()
+        self.leftTargetSpeed = 0
+        self.rightTargetSpeed = 0
+
+        # speeds are in RPM, same as velocity control mode
+
+    def setShooterAngle(self, angle: Rotation2d) -> None:
+        self.targetAngle = min(
+            max(
+                constants.kShooterMinAngle,
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py]: what should be taken into account here is that we may want to fudge beyond the range of these angles as a result of bad startup / weird encoder nonsense
I'd add the fudge with this

File: subsystems/shootersubsystem.py [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+    def setRightShootingMotorSpeed(self, rpm: int) -> None:
+        self.rightTargetSpeed = rpm + SmartDashboard.getNumber(
+            constants.kRightMotorFudgeKey, 0
+        )
+        self.rightShootingMotor.set(
+            NEOBrushless.ControlMode.Velocity,
+            rpm * constants.kShootingMotorRatio,
+        )
+
+    def getShooterAngle(self) -> Rotation2d:
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py:109]: add a function which is shooter angle from the encoder from the falcon

**Review by lmaxwell24 — CHANGES_REQUESTED** [review]
nitpicks, after I think should be good
I'll go and fix the kV stuff

**Review by lmaxwell24 — DISMISSED** [review]
later thing is putting a good chunk of the shooter commands into 1 file but eh

File: constants.py [hunk 24]
```diff
@@ -685,3 +686,74 @@
...
+
+kShooterManualModeKey = "ShooterManualMode"
+kShooterAngleFudgeKey = "ShooterAngleFudge"
+kLeftMotorFudgeKey = "LeftMotorFudge"
+kRightMotorFudgeKey = "RightMotorFudge"
+
+kShootingMotorFudgeAmount = 50
+
+# radians
+kShootingAngleFudgeAmount = 0.05
```
Comment (LandonBayer) [inline constants.py]: .05 rad = 2.9 degrees, seems like a big adjustment so while we're tuning we probably should make this lower

File: subsystems/shootersubsystem.py [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+            * constants.kAngleMotorRatio,
+        )
+
+    def setLeftShootingMotorSpeed(self, rpm: int) -> None:
+        self.leftTargetSpeed = rpm + SmartDashboard.getNumber(
+            constants.kLeftMotorFudgeKey, 0
+        )
+        self.leftShootingMotor.set(
+            NEOBrushless.ControlMode.Velocity,
+            rpm * constants.kShootingMotorRatio,
```
Comment (LandonBayer) [inline subsystems/shootersubsystem.py]: Why is rpm the input here instead of targetSpeed?

File: subsystems/shootersubsystem.py [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+        )
+
+    def setRightShootingMotorSpeed(self, rpm: int) -> None:
+        self.rightTargetSpeed = rpm + SmartDashboard.getNumber(
+            constants.kRightMotorFudgeKey, 0
+        )
+        self.rightShootingMotor.set(
+            NEOBrushless.ControlMode.Velocity,
+            rpm * constants.kShootingMotorRatio,
```
Comment (LandonBayer) [inline subsystems/shootersubsystem.py]: Same as above

File: subsystems/shootersubsystem.py [hunk 27]
```diff
@@ -0,0 +1,174 @@
...
+            abs(self.rightTargetSpeed - self.getRightShooterSpeed())
+            < constants.kShooterSpeedTolerance
+        )
+
+    def readyToShoot(self) -> bool:
+        return (
+            self.angleOnTarget()
+            and self.leftMotorSpeedOnTarget()
+            and self.rightMotorSpeedOnTarget()
+            and SmartDashboard.getBoolean(constants.kRobotAngleOnTargetKey, False)
```
Comment (LandonBayer) [inline subsystems/shootersubsystem.py:150]: Why is this here if you already check the function?

File: commands/shooter/decreaseleftmotorspeed.py [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+
+    def initialize(self) -> None:
+        print(f"Command: {self.getName()}")
+
+    def execute(self) -> None:
+        SmartDashboard.putNumber(
+            constants.kLeftMotorFudgeKey,
+            SmartDashboard.getNumber(constants.kLeftMotorFudgeKey, 0)
```
Comment (LandonBayer) [inline commands/shooter/decreaseleftmotorspeed.py:18]: How does this actually change it

**Review by LandonBayer — CHANGES_REQUESTED** [review]

File: constants.py (hunk 24 above)
Comment (icai0) [reply constants.py]: can be changed later

**Review by LandonBayer — DISMISSED** [review]

**Review by lmaxwell24 — DISMISSED** [review]
